### PR TITLE
Transaction system improvements and docs

### PR DESCRIPTION
**Pull request by astubbs**
Clarifies transaction system with much better documentation.
Fixes a potential race condition which could cause a offset leaks between transactions.
Introduces lock acquisition time outs.
Fixes a potential issue with removing records from the retry incorrectly from having an inconsistency between compareTo and equals in the retry TreeMap.

- [x] Add bulk parallel transactions to features
- [x] change base branch to master
- [x] finish tests
- [x] document what happens upon timeouts - fail fast
- [x] changelog
- [x] consider cherry picking tangents 

Blocked by:
- #398
- #399
- #406

### CODE REVIEW

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java [hunk 1]
```diff
@@ -86,19 +86,62 @@ public enum ProcessingOrder {
...
+         * a) All records produced from a given source offset will be visible, or none will be
+         * ({@link org.apache.kafka.common.IsolationLevel#READ_COMMITTED}).
+         * <p>
+         * b) If any records making up a transaction have a terminal issue being produced, or the system crashes before
+         * finishing sending all the records and committing, none will ever be visible and the system will retry th e
+         * whole set in a new transaction.
+         * <p>
+         * c) a source offset, and it's produced records will be committed as a set. Normally: either the record
+         * producing could fail, or the committing of the source offset could fail, as they are individual operations.
+         * When uUsing Transactions, they are committed together - so if either operations fail, the transaction will
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java:104]: Using

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java [hunk 2]
```diff
@@ -86,19 +86,62 @@ public enum ProcessingOrder {
...
+         * never get committed, and upon recovery the system will retry the set again (and no duplicates will be visible
+         * in the topic).
+         * <p>
+         * Slowest of the options, but no duplicates in Kafka caused by producing a record from consuming a record, that
+         * has already had its offset committed (message replay may cause duplicates in external systems which is
+         * unavoidable).
+         * <p>
+         * Note that the system uses very large transactions (for transaction standards), by only committing by default
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java:112]: Ah - this will cause massive stalls in reading the topic - min latency consumption of 5 seconds. System needs to reduce the frequency as does ks. Say default same asks? And encourage configuring.

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java (hunk 2 above)
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java:112]: Highlight this as a disadvantage of using transactions for consumers at the head of the topic

File: src/docs/README_TEMPLATE.adoc [hunk 3]
```diff
@@ -745,6 +745,12 @@ CAUTION: This cannot be true for any externally integrated third party system, u
 
 For implementations details, see the <<Transactional System Architecture>> section.
 
+.From the Options Javadoc
+[source,java,indent=0]
+----
+include::{project_root}/parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java[tag=transactionalJavadoc]
```
Comment (astubbs) [inline src/docs/README_TEMPLATE.adoc:751]: Is it possible to do a better render than the raw java doc in the ReadMe?

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java [hunk 4]
```diff
@@ -86,19 +86,62 @@ public enum ProcessingOrder {
...
+         * committing - that being upon failure, all records not previously committed will be replayed. Reducing this
+         * configuration places higher loads on the broker, but will reduce (but cannot eliminate) replay upon failure.
+         * <p>
+         * When producing multiple records (see {@link ParallelStreamProcessor#pollAndProduceMany}), all records must
+         * have been produced successfully to the broker before the transaction will commit, after which all will be
+         * visible together, or none.
+         * <p>
+         * Also records produced while running in this mode, won't be seen by consumer running in
+         * {@link ConsumerConfig#ISOLATION_LEVEL_CONFIG} {@link org.apache.kafka.common.IsolationLevel#READ_COMMITTED}
+         * mode until the transaction is complete and all records are produced successfully. Records produced into a
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java:124]: All of nothing

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/AbstractParallelEoSStreamProcessorTestBase.java [hunk 5]
```diff
@@ -47,6 +47,10 @@
 import static org.mockito.Mockito.*;
 import static pl.tlinkowski.unij.api.UniLists.of;
 
+/**
+ * @author Antony Stubbs
+ * @see AbstractParallelEoSStreamProcessor
+ */
 // todo migrate commit assertion methods in to a Truth Subject
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/AbstractParallelEoSStreamProcessorTestBase.java:54]: improvements/deprecate-old-commit-asserts

File: src/docs/README_TEMPLATE.adoc (hunk 3 above)
Comment (astubbs) [reply src/docs/README_TEMPLATE.adoc:753]: not for this version - maybe later

File: parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/ParallelConsumerOptionsTest.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * @author Antony Stubbs
+ */
+class ParallelConsumerOptionsTest {
+
+    /**
+     * Test the deprecation phase of commit frequency
+     */
+    @Test
+    void setTimeBetweenCommits() {
+        var newFreq = Duration.ofMillis(100);
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/ParallelConsumerOptionsTest.java:21]: ensure system works

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java [hunk 7]
```diff
@@ -86,19 +87,71 @@ public enum ProcessingOrder {
...
 
+        // tag::transactionalJavadoc[]
         /**
-         * Periodically commits through the Producer using transactions. Slowest of the options, but no duplicates in
-         * Kafka guaranteed (message replay may cause duplicates in external systems which is unavoidable with Kafka).
+         * Periodically commits through the Producer using transactions.
          * <p>
-         * This is separate from using an IDEMPOTENT Producer, which can be used, along with {@link
-         * CommitMode#PERIODIC_CONSUMER_SYNC} or {@link CommitMode#PERIODIC_CONSUMER_ASYNCHRONOUS}.
+         * The benefits of using this mode are:
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java:94]: Make sure to clearly explain that unlike normal use of the raw clients, parallel consumer will mix together records from different concurrent operations into the single transaction as the producer in science is shared between current work. If you would like support for separate transactions per processing action please file an issue.

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractOffsetCommitter.java [hunk 8]
```diff
@@ -51,6 +51,7 @@ protected void postCommit() {
     }
 
     protected void preAcquireWork() {
+//        this.
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractOffsetCommitter.java:54]: delete

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/internal/ProducerManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+            assertThat(pm).transactionNotOpen();
+            assertThat(pm).stateIs(COMMIT);
+        }
+    }
+
+    /**
+     * There's GOT to be a better way to test this corner case than this
+     */
+    @Test
+    void producedRecordsCantBeInTransactionWithoutItsOffset() {
```
Comment (astubbs) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/internal/ProducerManagerTest.java:220]: Is it possible to test this using just the control loop function?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            if (transactionManager == null) {
+                return false;
+            } else {
+                return transactionManager.isTransactional();
+            }
+        }
+    }
+
+    @SneakyThrows
+    private TransactionManager getTransactionManager() {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java:131]: with the state tracking is this still needed?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        if (txManagerField == null) return null;
+        TransactionManager transactionManager = (TransactionManager) txManagerField.get(producer);
+        return transactionManager;
+    }
+
+    /**
+     * TODO talk about alternatives to this brute force approach for retrying committing transactions
+     */
+    @SneakyThrows
+    protected boolean isTransactionCompleting() {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java:141]: still needed?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        if (producer instanceof MockProducer) return false;
+        return (boolean) txManagerMethodIsCompleting.invoke(getTransactionManager());
+    }
+
+    /**
+     * TODO talk about alternatives to this brute force approach for retrying committing transactions
+     */
+    @SneakyThrows
+    protected boolean isTransactionReady() {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java:150]: still needed?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/WorkContainer.java [hunk 13]
```diff
@@ -225,8 +227,24 @@ public boolean hasPreviouslyFailed() {
     }
 
     public boolean isAvailableToTakeAsWork() {
-        // todo missing boolean notAllowedMoreRecords = pm.isBlocked(topicPartition);
+        // todo missing boolean notAllowedMoreRecords = pm.isBlocked(topicPartition);?
         return isNotInFlight() && !isUserFunctionSucceeded() && hasDelayPassed();
     }
 
+    // todo delete
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/WorkContainer.java:234]: delete

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/WorkContainer.java [hunk 14]
```diff
@@ -225,8 +227,24 @@ public boolean hasPreviouslyFailed() {
...
+//    public void onPostAddToMailBox() {
+////        this.produceLock.map(locl->lock.finish());
+//    }
+
+    /**
+     * todo docs
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/WorkContainer.java:240]: docs

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/internal/ProducerManagerTest.java (hunk 9 above)
Comment (astubbs) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/internal/ProducerManagerTest.java:220]: try implementing with direct control loop

File: parallel-consumer-core/src/test/java/io/confluent/csid/utils/BlockedThreadAsserter.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.google.common.truth.Truth;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.awaitility.Awaitility.await;
+
+/**
+ * todo docs
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/csid/utils/BlockedThreadAsserter.java:11]: docs

File: parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/truth/ParallelEoSStreamProcessorSubject.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+     * Returns an assertion builder for a {@link ParallelEoSStreamProcessor} class.
+     */
+    @SubjectFactoryMethod
+    public static Factory<ParallelEoSStreamProcessorSubject, ParallelEoSStreamProcessor> parallelEoSStreamProcessors() {
+        return ParallelEoSStreamProcessorSubject::new;
+    }
+
+
+    public void hasCommittedToAnything(int offset) {
+        throw new InternalRuntimeError("");
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/truth/ParallelEoSStreamProcessorSubject.java:39]: not used?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 17]
```diff
@@ -676,13 +707,40 @@ private <R> void controlLoop(Function<PollContextInternal<K, V>, List<R>> userFu
...
         }
 
+
+        // acquiring the commit lock here, still allows user function side effects to run while committing a transaction
+        // todo optionally prevent side affects during transaction commit by acquiring the lock before getting work
+        /**
+         * call {@link ProducerManager#preAcquireWork()} early, to initiate the record sending barrier for this transaction
+         */
+        // todo also check tx is used
+        // could do this optimistically as well, and only get the lock if state is dirty
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java:717]: cleanup

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 18]
```diff
@@ -676,13 +707,40 @@ private <R> void controlLoop(Function<PollContextInternal<K, V>, List<R>> userFu
...
+        final boolean shouldTryCommitNow = isTimeToCommitNow();
+        updateLastCommitCheckTime();
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java:719]: why isn't this inside the checking function isTimeToCommitNow?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCModule.java [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.function.Supplier;
+
+/**
+ * DI
+ * <p>
+ * todo docs
+ * <p>
+ * A-la' Dagger.
+ *
+ * @author Antony Stubbs
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCModule.java:20]: docs

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCModuleProd.java [hunk 20]
```diff
@@ -0,0 +1,15 @@
+package io.confluent.parallelconsumer.internal;
+
+import io.confluent.parallelconsumer.ParallelConsumerOptions;
+
+/**
+ * todo docs
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCModuleProd.java:6]: remove?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.kafka.clients.producer.internals.TransactionManager;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.ProducerFencedException;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.Map;
+
+/**
+ * todo docs
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java:21]: docs

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/utils/RecordFactory.java [hunk 22]
```diff
@@ -0,0 +1,24 @@
+package io.confluent.parallelconsumer.integrationTests.utils;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * todo docs
```
Comment (astubbs) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/utils/RecordFactory.java:9]: docs

File: parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/truth/ConsumerRecordsSubject.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * Returns an assertion builder for a {@link ConsumerRecords} class.
+     */
+    @SubjectFactoryMethod
+    public static Factory<ConsumerRecordsSubject, ConsumerRecords> consumerRecordses() {
+        return ConsumerRecordsSubject::new;
+    }
+
+    public void containsOffset(int blockFreeRecords) {
+        failWithoutActual(Fact.simpleFact("not-implemented"));
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/truth/ConsumerRecordsSubject.java:35]: remove?

**Review by astubbs — COMMENTED** [review]
....

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/state/ModelUtils.java [hunk 24]
```diff
@@ -23,4 +37,46 @@ public WorkContainer<String, String> createWorkFor(long offset) {
...
+
+    @NonNull
+    private ConsumerRecord<String, String> createConsumerRecord(String topic) {
+        var cr = new ConsumerRecord<>(topic, 0, nextOffset, "a-key", "a-value");
+        nextOffset++;
+        return cr;
+    }
+
+    // todo duplicate?
+    public ProducerRecord<String, String> createProducerRecords() {
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/state/ModelUtils.java:73]: duplicate?

**Review by astubbs — COMMENTED** [review]
...

**Review by astubbs — COMMENTED** [review]
basics reviewed

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java [hunk 25]
```diff
@@ -38,12 +49,35 @@
...
      */
+    @Getter
     private ReentrantReadWriteLock producerTransactionLock;
 
+    /**
+     * todo docs
+     */
+    @ToString.Include
+    @Getter
+    private volatile ProducerState producerState = ProducerState.INSTANTIATED;
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:79]: move state into producer wrap?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java [hunk 26]
```diff
@@ -220,20 +373,39 @@ public void close(Duration timeout) {
...
             throw new ConcurrentModificationException(this.getClass().getSimpleName() + " is not safe for multi-threaded access");
         }
-        writeLock.lock();
+
+        // lock the lock
+        // todo remove or use sensible value
+        boolean gotLock = false;
+        try {
+            log.debug("Acquiring commit lock...");
+            gotLock = writeLock.tryLock() || writeLock.tryLock(6, TimeUnit.SECONDS);
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:400]: timeout

**Review by astubbs — COMMENTED** [review]
only timeouts and exceptions left

**Review by astubbs — COMMENTED** [review]
LGTM

**Review by astubbs — COMMENTED** [review]
...

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionTimeoutsTest.java [hunk 27]
```diff
@@ -0,0 +1,213 @@
...
+            long offset = recordContexts.offset();
+            if (offset == offsetToFail) {
+                // triggers deadlock as controller can't acquire commit lock fast enough due to this sleeping thread
+                ThreadUtils.sleepQuietly(1000 * multiple);
+            }
+            return new ProducerRecord<>(getTopic() + "-output", "random");
+        });
+
+
+        // send, process, commit
```
Comment (astubbs) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionTimeoutsTest.java:109]: add sanity checks for early messages before abort tests

File: parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/internal/ProducerManagerTest.java [hunk 28]
```diff
@@ -275,39 +300,54 @@ void producedRecordsCantBeInTransactionWithoutItsOffsetDirect() {
...
+                log.debug(msg);
+                await(msg).untilAsserted(() -> assertThat(pc.getWorkMailBox()).hasSize(1));
+            }
 
             // won't block - not dirty
-            pc.controlLoop(userFunc, o -> {
-            });
+            // collects first work results
+            // todo delete?
+//            pc.controlLoop(userFunc, o -> {
```
Comment (astubbs) [inline parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/internal/ProducerManagerTest.java:325]: delete

**Review by astubbs — COMMENTED** [review]
still need to finish of the extra test parts

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionAndCommitModeTest.java [hunk 29]
```diff
@@ -111,7 +113,7 @@ public void testHighMaxPollEnum(CommitMode commitMode, ProcessingOrder order) {
 
     private void runTest(int maxPoll, CommitMode commitMode, ProcessingOrder order) {
         //        int expectedMessageCount = 50_000;
```
Comment (rkolesnev) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionAndCommitModeTest.java:115]:
```suggestion
```

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionAndCommitModeTest.java [hunk 30]
```diff
@@ -111,7 +113,7 @@ public void testHighMaxPollEnum(CommitMode commitMode, ProcessingOrder order) {
...
-        int expectedMessageCount = 10_000;
+        int expectedMessageCount = 30_000;
 //        int expectedMessageCount = 10_000;
 //        int expectedMessageCount = 1_000;
```
Comment (rkolesnev) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionAndCommitModeTest.java:118]:
```suggestion
```

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionTimeoutsTest.java [hunk 31]
```diff
@@ -0,0 +1,221 @@
...
+
+        //// send, process, commit
+        // assert output topic
+        assertConsumedOffset(5); // happy path, all base records committed ok
+
+        //// send, process, block, retry, succeed
+        // send 3 more records
+        getKcu().produceMessages(getTopic(), 3);
+
+        // assert output topic - has got the new records due to commit blocked
```
Comment (rkolesnev) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionTimeoutsTest.java:210]:
```suggestion
        // assert output topic - ensuring it hasn't got the new records due to commit blocked
```

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionTimeoutsTest.java [hunk 32]
```diff
@@ -0,0 +1,221 @@
...
+        assertConsumedOffset(5); // happy path, all base records committed ok
```
Comment (rkolesnev) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/integrationTests/TransactionTimeoutsTest.java:211]:
```suggestion
        assertConsumedOffset(5); // only base records
```

**Review by rkolesnev — APPROVED** [review]
Overall LGTM - few suggestions added.

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java [hunk 33]
```diff
@@ -110,6 +177,64 @@ public enum CommitMode {
...
+     * Disabled by default.
+     * <p>
+     * Doesn't interfere with the transaction.
+     * <p>
+     * Recommended to leave this off to avoid side effect duplicates upon rebalance after a crash. Enabling could
+     * improve performance as the produce lock will only be taken right before it's needed to produce the result record,
+     * instead of pessimistically.
+     */
+    @Builder.Default
+    private boolean allowEagerProcessingDuringTransactionCommit = false;
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelConsumerOptions.java:206]: I've included it as a reference in the javadoc for transactions - it's an advanced setting that I think should be left false

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java [hunk 34]
```diff
@@ -34,23 +40,40 @@
...
 
-    protected final ProducerWrapper<K, V> producer;
+    @Getter
+    protected final ProducerWrapper<K, V> producerWrapper;
 
     private final ParallelConsumerOptions<K, V> options;
 
     /**
-     * The {@link KafkaProducer} isn't actually completely thread safe, at least when using it transactionally. We must
+     * The {@link KafkaProducer} isn't actually completely thread safe, at least when using it transitionally. We must
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:50]: transactionally - if not using transactions, I didn't find this problem

[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:50]: ah - whoops! good find

**Review by astubbs — COMMENTED** [review]
...

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java [hunk 35]
```diff
@@ -262,7 +269,7 @@ protected void commitOffsets(@NonNull Map<TopicPartition, OffsetAndMetadata> off
                             // try wait again
                             commitTransaction();
                         }
-                        boolean transactionModeIsReady = lastErrorSavedForRethrow == null || !lastErrorSavedForRethrow.getMessage().contains("Invalid transition attempted from state READY to state COMMITTING_TRANSACTION");
+                        boolean transactionModeIsReady = lastErrorSavedForRethrow == null || !lastErrorSavedForRethrow.getMessage().contains("Invalid transaction attempted from state READY to state COMMITTING_TRANSACTION");
```
Comment (rkolesnev) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:272]: Looks like it was meant to say 
```suggestion
                        boolean transactionModeIsReady = lastErrorSavedForRethrow == null || !lastErrorSavedForRethrow.getMessage().contains("Invalid transition attempted from state READY to state COMMITTING_TRANSACTION");
```

Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java:272]: 🤦

**Comment by rkolesnev** [comment]
LGTM - already approved on previous review as suggestions were not fundamental.
